### ISSUE TITLE
notification endpoints cannot be configured with environment variables

### ISSUE THREAD

**Issue by diptanu** [distribution/distribution#625]
We are using the notification API to build an index of images pushed for building a search API. We have a configuration like -

```
notifications:
    endpoints:
        - name: pandora
          url: http://our-event-server.orgname.net/events
          timeout: 1s
          threshold: 10
          backoff: 1s
          disabled: false
```

Can we do something like REGISTRY_NOTIFICATIONS_ENDPOINTS_NAME_URL=<the_url_here>
Since the endpoints is a list, I am not really sure what is the convention for defining environment variables are for this section.

**Comment by stevvooe** [comment]
I am not sure the environment variables are the best approach here. There isn't an obvious mapping between listed items in the config and an unrolled environment.

We should probably just allow some sort of environment pointer.

@BrianBland Any thoughts here?

**Comment by diptanu** [comment]
@stevvooe @BrianBland It would be great if I can use some environment variable as a pointer to the value of a key in the yaml config.

**Comment by diptanu** [comment]
@stevvooe Is there a work around here?? We basically want to use different notification endpoint URLs for different AWS Regions.

**Comment by stevvooe** [comment]
@diptanu I'd recommend you attempt to fix the configuration system to address this with an env pointer for any value. It should be a simple fix but we cannot prioritize it.

Save that, given the guarantees of the notification system, it would probably be best to run a local process to accept notifications and then forward them to the more dynamic endpoint location.

**Comment by diptanu** [comment]
@stevvooe Thanks! Let me see what I can do.

**Comment by stevvooe** [comment]
@diptanu Great! Let me know if you need anything.

**Comment by aleksnikolic** [comment]
Hi guys, I am having the same problem here. Trying to use portus as a token authentication service but don't know how to override notification config. 

Similar like this:

notifications:
  endpoints:
    - name: portus
      url: http://portus.test.lan/v2/webhooks/events
      timeout: 500ms
      threshold: 5
      backoff: 1s

Any idea how to override it during docker run -e registry:2.0?

**Comment by stevvooe** [comment]
@aleksnikolic This issue identifies an existing gap in functionality. You'll need to replace the configuration file.

We probably want to add an ENV placeholder for configuration values. @diptanu Are you planning on taking a crack at this?

**Comment by lkoenigsberger** [comment]
Any news on that topic?

**Comment by stevvooe** [comment]
@lkoenigsberger We don't really have this on our road map. If someone wants to take this on, they're more than welcome.

**Comment by clarktlaugh** [comment]
I just ran into this  It's a bit confusing, I think, in that you have a subset of the configuration options that are "special" and can't be configured like everything else can when running using the pre-built container.  I couldn't find this documented anywhere and spent many hours trying to figure out why it wasn't working.

Perhaps implementing support EVs for configuring a **single** notification target is a suitable solution which will meet the needs of most people?

But for now, What do I use as the base for the configuration file?  I want to ensure that I don't inadvertently change something else by not using the correct starting point.  Is it safe to create a new file and only specify the options I was previously using EVs for?

**Comment by stevvooe** [comment]
@clarktlaugh Sorry about the confusion. A lot of the details of how this works are covered here: https://github.com/docker/distribution/blob/master/docs/configuration.md#override-specific-configuration-options.

I would suggest creating a configuration file and build it out from there. I usually test with something like this:

```
version: 0.1
log:
  level: debug
  fields:
    service: registry
    environment: development
storage:
  filesystem:
    rootdirectory: /usr/local/var/lib/registry
http:
  addr: :5000
  debug:
      addr: localhost:5001
  headers:
      X-Content-Type-Options: [nosniff]
```

From there, features can be easily added. This file can be distributed via orchestration and bind mounted at runtime.

Given the [guarantees of the notification system](https://github.com/docker/distribution/blob/master/docs/notifications.md#considerations), I strongly suggest that the endpoints are local to the registry node. Treat this more like syslog, where the service is always available on a well-known port, rather than powered through service discovery. We've found this restriction to be immensely reliable in production. If you need to proxy events more dynamically, you can have that manifold do more complex routing and even be environment configurable or via a database. With this approach, we've successfully integrated with myriad queueing systems without touching the registry or its configuration at all.

---
Background:

It's not that these fields are special. Basically, the request of mirroring environment level configuration to yaml requires a consistent and usable linearization of the tree represented by yaml. Actually implementing something like this is much harder than it sounds.

Take this case:

```yaml
notifications:
  endpoints:
    - name: alistener
    # listener fields
```

`REGISTRY_NOTIFICATIONS_ENDPOINTS_ALISTENER_XXX` seems to be a top suggestion. However, to maintain compatibility, the env mapper needs to dip into the target configuration, know which field to map (in this case "name"). If we adopt this for just this type, adding new list types would create inconsistencies in expected behavior. The other issue is that we'd have ambiguity when targeting a list, since named entities would have to be unique. The correct approach would be to use a map type here, which would enforce uniqueness.

The consistent approach would be something like `REGISTRY_NOTIFICATIONS_ENDPOINTS_0_XXX`, which would let one index list entries. This is also problematic, since now we have to know which entry is which.

Another option is to store yaml in the actual environment variable and interpret it at runtime. This has its own issues. (We may actually support this, so give it a try @RichardScothern @aaronlehmann).

A better approach, favored by the v1 registry, is to actually call out certain fields as being a part of the environment. This way, the notification definition would look as follows:

```yaml
notifications:
  endpoints:
    - name: alistener
      url: _env(REGISTRY_NOTIFICATION_ALISTENER_URL)
    # listener fields
```

This let's us affix parts of the tree with a handle. Unfortunately, this means every string field becomes part of a micro-format, which has problems of its own.

That said, I'd like to see a PR that takes the `REGISTRY_NOTIFICATIONS_ENDPOINTS_ALISTENER_XXX` approach to see how bad it really is. At least, we could address this use case and get this bug closed out.

**Comment by aaronlehmann** [comment]
> Another option is to store yaml in the actual environment variable and interpret it at runtime.

Yes, this is already supported.

**Comment by sshipway** [comment]
A +1 for allowing at least one notification to be defined using Environment Variables.  There are situations where it is not practical to create a volume mount but passing env vars works.

@aaronlehmann - how do you pass YAML in the environment?  I have not found any documentation explaining how to do that.

**Comment by aaronlehmann** [comment]
Environment value variables are parsed as YAML. If you are specifying something whose value should be an array or structure, you can give the value as YAML.

**Comment by sshipway** [comment]
So, does that mean that this would work (if the \n are expanded) --
````
-e REGISTRY_NOTIFICATIONS_ENDPOINTS="- name: foo\n  url: http://foosvr/event\n"
````
If so, then this is a good workaround.  Spaces might be a bit sensitive.

**Comment by broegner** [comment]
Is there any possibility to get the _env() solution working with V2 or is there maybe existing another way to have a pointer in V2?

**Comment by EugenMayer** [comment]
Since its yet not clear if things do work out and i found this issue searching for something else - it does work out setting it using the ENV variables 

https://github.com/EugenMayer/docker-rancher-extra-catalogs/blob/master/templates/registry-slim/12/docker-compose.yml#L60

Or like in this concrete setup: https://github.com/EugenMayer/docker-image-portus/blob/master/test/docker-compose.yml#L68

**Comment by tboerger** [comment]
What about taking the `REGISTRY_NOTIFICATIONS_ENDPOINTS_*` variables like arrays/slices? [confd](https://github.com/kelseyhightower/confd) is doing something similar with environment variables. Than you should be able to do something like this:

```
export REGISTRY_NOTIFICATIONS_ENDPOINTS_0_NAME=endpoint1
export REGISTRY_NOTIFICATIONS_ENDPOINTS_0_DISABLED=false
export REGISTRY_NOTIFICATIONS_ENDPOINTS_0_URL=http://localhost:8080/api/hook

export REGISTRY_NOTIFICATIONS_ENDPOINTS_1_NAME=endpoint2
export REGISTRY_NOTIFICATIONS_ENDPOINTS_1_DISABLED=false
export REGISTRY_NOTIFICATIONS_ENDPOINTS_1_URL=http://localhost:9000/api/hook
```

**Comment by 374990191ls** [comment]
FYI

```sh
docker run -dt --name "test_notification" \
 -p 6500:5000 \
 -v /mnt/registry-storage:/var/lib/registry \
 -e REGISTRY_LOG_LEVEL=warn \
 -e REGISTRY_STORAGE_DELETE_ENABLED=True \
 -e REGISTRY_HTTP_SECRET=stubb-belle-seraglio \
 -e REGISTRY_NOTIFICATIONS_ENDPOINTS="- name: alistener
   url: http://examples.com/notifications
   timeout: 500
   threshold: 5
   backoff: 1
   " \
 registry:latest
```

**Comment by stevvooe** [comment]
@tboerger @374990191ls Do you have any recommendation on where we should put this in the docs?

Also, in your opinion, would following proposal above (having a default `_env` in the image), be a better option? Should we pursue that further?

@mstanleyjones PTAL

**Comment by tboerger** [comment]
If the confd-like behavior is not possible something like `_env(REGISTRY_NOTIFICATION_ALISTENER_URL)` would already help.

**Comment by lifeisfoo** [comment]
I've just adapted the @sshipway solution to a `docker-compose.yml` file, and this works nicely:

```
version: '3.4'

services:
  registry:
    image: registry:2
    environment:
      REGISTRY_NOTIFICATIONS_ENDPOINTS: "- name: notifications-test\n  url: http://my-notification-server:8080\n  timeout: 5s\n  threshold: 8\n  backoff: 10s"
  
```

**Comment by josegomezr** [comment]
I ran into this problem too.

I wanted to configure an Storage Middleware and only could do it with @lifeisfoo solution.

In spirit of trying to find a solution, I did write some Go Code (taken that I'm not a native go developer) and with my naive approach I could make what @tboerger  suggested work.

> ```
> export REGISTRY_NOTIFICATIONS_ENDPOINTS_0_NAME=endpoint1
> ```

However I don't quite know how to contribute this, or if there's a pending discussion or if it's even good enough Go code 😂

Code [is here](https://github.com/distribution/distribution/compare/main...josegomezr:add_env_support_for_arrays?expand=1)
